File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int dp[505][505];

int solution(vector<vector<int>> triangle) {
    
    int n = triangle.size();
    
    dp[0][0] = triangle[0][0];
    
    for(int i = 1; i < n; i++){
        dp[i][0] = dp[i - 1][0] + triangle[i][0];
        
        for(int j = 1; j <= i; j++){
            dp[i][j] = max(dp[i - 1][j - 1], dp[i - 1][j]) + triangle[i][j];
        }
    }
    
    int answer = *max_element(dp[n - 1], dp[n - 1] + n + 1);
    
    return answer;
}

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.java
import java.util.*;

class Solution {
    static int[][] dp;
    
    public int solution(int[][] triangle) {
        int answer = 0;
        
        int N = triangle.length;
        
        dp = new int[N][N];
        dp[0][0] = triangle[0][0]; // base condition
        
        for(int i = 1; i < N; i++){
            int M = triangle[i].length;
            
            for(int j = 0; j < M; j++){
                if(j - 1 >= 0) dp[i][j] = dp[i - 1][j - 1] + triangle[i][j];
                dp[i][j] = Math.max(dp[i][j], dp[i - 1][j] + triangle[i][j]);
            }
        }
        
        for(int i = 0; i < N; i++) {
            answer = Math.max(answer, dp[N - 1][i]);
        }
        
        return answer;
    }
}